src/System.CommandLine/Invocation/CliCommandAction.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.CommandLine.Invocation;
+
+// FIX: separate files
+
+public abstract class CliAction
+{
+    private protected CliAction()
+    {
+    }
+
+    // FIX: (CliAction) change this to accommodate the fact that the core action isn't "exclusive" either, and some actions might eventually be created that run after it.
+    public bool Exclusive { get; protected set; } = true;
+}
+
+/// <summary>
+/// Defines the behavior of a symbol.
+/// </summary>
+public abstract class SynchronousCliAction : CliAction
+{
+    /// <summary>
+    /// Performs an action when the associated symbol is invoked on the command line.
+    /// </summary>
+    /// <param name="parseResult">Provides the parse results.</param>
+    /// <returns>A value that can be used as the exit code for the process.</returns>
+    public abstract int Invoke(ParseResult parseResult);
+}
+
+public abstract class AsynchronousCliAction : CliAction
+{
+    /// <summary>
+    /// Performs an action when the associated symbol is invoked on the command line.
+    /// </summary>
+    /// <param name="parseResult">Provides the parse results.</param>
+    /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+    /// <returns>A value that can be used as the exit code for the process.</returns>
+    public abstract Task<int> InvokeAsync(ParseResult parseResult, CancellationToken cancellationToken = default);
+}
+
+public class SynchronousCliCommandAction : SynchronousCliAction
```

Comment:
Maybe I do not understand the usage, but it does not look like folks can create their own actions, and I recalled that as a key use case.

Reply:
> what do we gain by exposing this type?

> also: by naming it Command we limit the usage to commands only, while we may want to use it for directive or active options one day too

This is the goal -- to enable people to distinguish between command actions and other kinds of actions, which tend to represent interception of some kind, whether by a directive, help, or a parse error.

Reply:
> Maybe I do not understand the usage, but it does not look like folks can create their own actions, and I recalled that as a key use case.

They can. There are a number of examples in the tests.

Reply:
> This is the goal -- to enable people to distinguish between command actions and other kinds of actions, which tend to represent interception of some kind, whether by a directive, help, or a parse error.

Just thinking out loud: maybe we should consider exposing `Symbol Origin` in the `CliAction` to allow for such checks? 

Then we would need to expose fewer public types and the checks in theory would be simpler (knowing about the existance of both `CliCommandAction` types)?

```diff
- bool isCommandAction = parseResult.Action is SynchronousCliCommandAction or AsynchronousCliAction;
+ bool isCommandAction = parseResult.Action.Origin is Command;
```

The tricky thing would be choosing the right symbol for `ParseErrorAction` (it could be always root command, just null or maybe the symbol that produced errors if it was the only one with parse issues)

PS. Origin may be a bad name, I hope you get my intentions.

Reply:
> The tricky thing would be choosing the right symbol for ParseErrorAction

A `ParseErrorAction` can't reliably be associated with a specific symbol.

Reply:
Part of why we have the extra types here is that the previous concept of an "anonymous" `CliAction` has been merged with the `CliCommandAction`, which is probably the wrong thing to do. The anonymous pattern (used by the `SetHandler` methods to create a `CliAction` instance whose implementation is injected via a delegate) shouldn't be coupled to whether the action is related to a command.

I can remove the concept of a command action. The check for whether it's the intended action can be satisfied like this:

```csharp
if (parseResult.Action != someCommand.Action)
```

Pattern matches will have to based on very specific types rather than base types.

Anonymous action types (as internal types) will probably need to be put back in place.

I'll make these changes and we can see how it looks.